### PR TITLE
rewriting: Move dce to rewriter

### PR DESCRIPTION
**Pull request by alexarice**
In mlir, dead code elimination seems to be part of the rewriter itself. I am planning to write a full (region-level) dead code elimination pass but as a start I have moved the "trivial" dead code elimination to the rewriter. I think this change will make more sense with the full pass but I thought this was already enough for a PR.

This also makes canonicalization a little bit more aggressive as dead code elim and canonicalization patterns can be interleaved. This has changed one of the stencil tests so it would be good to get a confirmation that this is OK.

Looks like a lot of changes but it's mostly moving code around which git hasn't managed to keep track of.

### CODE REVIEW

File: xdsl/pattern_rewriter.py [hunk 1]
```diff
@@ -650,7 +660,7 @@ class PatternRewriteWalker:
     The walker will also walk recursively on the created operations.
     """
 
-    pattern: RewritePattern
+    pattern: RewritePattern = field(default_factory=NoRewrite)
```
Comment (superlopuh) [inline xdsl/pattern_rewriter.py:663]: Why this change?

File: xdsl/transforms/dead_code.py
Comment (superlopuh) [inline xdsl/transforms/dead_code.py:1]: Why a new file?

Comment (alexarice) [reply xdsl/transforms/dead_code.py:1]: Import loops

File: xdsl/pattern_rewriter.py (hunk 1 above)
Comment (alexarice) [reply xdsl/pattern_rewriter.py:663]: It made the toy example a bit neater, but it could be dropped

Comment (superlopuh) [reply xdsl/pattern_rewriter.py:663]: My gut says drop it FWIW

File: xdsl/transforms/dead_code.py
Comment (superlopuh) [reply xdsl/transforms/dead_code.py:1]: If it's just the PatternRewriteWalker I would add the import inside the dce function and call it a day, it feels a little messy to split this file

**Comment by alexarice** [comment]
I think a better approach to doing this will be to keep dce separate from the rewrite machinery, but add an option on the pattern rewriter to perform a module pass at the end of each rewrite phase. This allows the same functionality as the mlir version but is a bit cleaner in my opinion and much more modular.

I'm going to close this PR in favor of doing the above.
